includes/restoreMappings.js

```diff
@@ -0,0 +1,112 @@
+
+class Restore {
+  constructor(db) {
+    this.db = db;
+    this.totalDocsRestored = 0;
+    this.batchCounter = 0;
+  }
+
+  /**
+   * Mapper for converting a backup file line to an array of documents pending restoration.
+   *
+   * @param {object} backupLine object representation of a backup file line {lineNumber: #, line: '...'}
+   * @returns {array} array of documents parsed from the line
+   */
+  backupLineToDocsArray = async(backupLine) => {
+    if (backupLine && backupLine.line !== '') {
+      // see if it parses as JSON
+      try {
+        const arr = JSON.parse(backupLine.line);
+        // if it's an array with a length
+        if (Array.isArray(arr) && arr.length > 0) {
+          return arr;
+        } else {
+          return Promise.reject(new error.BackupError('BackupFileJsonError', `Error on line ${backupLine.lineNumber} of backup file - not an array`));
+        }
+      } catch (e) {
+        // If the line can't be parsed as JSON it is most likely an incomplete write.
+        // If the backup was resumed we can ignore the error because the line will be repeated.
+        const parseError = new error.BackupError('BackupFileJsonError', `Error on line ${backupLine.lineNumber} of backup file - cannot parse as JSON`);
+        // If the backup wasn't resumed then it is invalid and we should error, but we have no way to detect this atm.
+        mappingDebug(`${parseError}`);
+      }
+    }
+    // Return an empty array if there was a blank or ignorable line
+    return [];
+  };
+
+  /**
+   * Mapper to wrap an array of docs in batch metadata
+   * @param {array} docs an array of documents to be restored
+   * @returns {object} a pending restore batch {batch: #, docs: [...]}
+   */
+  docsToRestoreBatch = async(docs) => {
+    return { batch: this.batchCounter++, docs };
+  };
+
+  /**
+   * Mapper for converting a pending restore batch to a _bulk_docs request
+   * and awaiting the response and finally returing a "restored" object
+   * with the number of restored docs and the running total.
+   *
+   * @param {object} restoreBatch a pending restore batch {batch: #, docs: [{_id: id, ...}, ...]}
+   * @returns {object} a restored batch object {documents: #, total: #}
+   */
+  pendingToRestored = async(restoreBatch) => {
+    // Save the batch number
+    const batch = restoreBatch.batch;
+    mappingDebug(`Preparing to restore ${batch}`);
+    // Remove it from the restoreBatch since we'll use that as our payload
+    delete restoreBatch.batch;
+    if (restoreBatch.docs && restoreBatch.docs.length > 0) {
+      mappingDebug(`Restoring batch ${batch} with ${restoreBatch.docs.length} docs.`);
+      // if we are restoring known revisions, we need to supply new_edits=false
+      if (restoreBatch.docs[0] && restoreBatch.docs[0]._rev) {
+        restoreBatch.new_edits = false;
+        mappingDebug('Using new_edits false mode.');
+      }
+      return await this.db.service.postBulkDocs({
```

Comment:
This looks... wrong? Await returns fulfillment value from `postBulkDocs` processing so we can then wrap it into promise again, since our func is `async`?

Wouldn't this to be simpler as `response = await this.db.service.postBulkDocs` in `try {} catch` instead?